src/com/ms/silverking/cloud/dht/daemon/storage/protocol/BaseRetrievalOperation.java

```diff
@@ -11,62 +11,59 @@
     }
-        
-    /*
-     * The below processInitialMessageGroupEntry is for a typical receive protocol
-     * where we only need to receive from a single replica.
-     * 
-     * Think about moving this into a subclass that protocols pull from if they implement this behavior.
-     */
-    @Override
-    public void processInitialMessageGroupEntry(DHTKey key, 
-            List<IPAndPort> primaryReplicas, List<IPAndPort> secondaryReplicas,
-            OpVirtualCommunicator<DHTKey, RetrievalResult> rvComm) {
-        BaseRetrievalEntryState  entryState;
-        
+
+    // Note that we do not handle local retrievals here - even though we could -
+    // because we want to group all local retrievals so that we can
+    // handle them en masse and incur only a single lock acquisition.
+    // By "forwarding" them, they will be handled in bulk
+
+    entryState = initializeEntryState(key, primaryReplicas, secondaryReplicas);
+    if (forwardingMode.forwards()) {
+      //initializeEntryState(entry, primaryReplicas, secondaryReplicas);
+      if (retrievalOperationContainer.containsLocalReplica(
+          primaryReplicas) || retrievalOperationContainer.containsLocalReplica(secondaryReplicas)) {
         if (debug) {
-            System.out.println("forwardingMode: "+ forwardingMode);
-            System.out.printf("p %s s %s\n", primaryReplicas, secondaryReplicas);
+          System.out.println("local forward");
         }
-        
-        // Note that we do not handle local retrievals here - even though we could - 
-        // because we want to group all local retrievals so that we can
-        // handle them en masse and incur only a single lock acquisition.
-        // By "forwarding" them, they will be handled in bulk 
-        
-        entryState = initializeEntryState(key, primaryReplicas, secondaryReplicas);
-        if (forwardingMode.forwards()) {
-            //initializeEntryState(entry, primaryReplicas, secondaryReplicas);
-            if (retrievalOperationContainer.containsLocalReplica(primaryReplicas) 
-                    || retrievalOperationContainer.containsLocalReplica(secondaryReplicas)) {
-                if (debug) {
-                    System.out.println("local forward");
-                }
-                rvComm.forwardEntry(operationContainer.localIPAndPort(), key);
-            } else {
-                try {
-                    rvComm.forwardEntry(entryState.getInitialReplica(), key);
-                } catch (NoSuchElementException nsee) {
-                    throw new RuntimeException("No replicas for: "+ key);
-                }
-            }
-        } else {
-            rvComm.forwardEntry(operationContainer.localIPAndPort(), key);
+        rvComm.forwardEntry(operationContainer.localIPAndPort(), key);
+      } else {
+        try {
+          rvComm.forwardEntry(entryState.getInitialReplica(), key);
+        } catch (NoSuchElementException nsee) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)